contrib/mesos/pkg/minion/task.go

```diff
@@ -0,0 +1,519 @@
+	bin          string                // required: path to executable
+	args         []string              // optional: process arguments
+	env          []string              // optional: process environment override
+	createLogger func() io.WriteCloser // factory func that builds a log writer
+	cmd          systemProcess         // process that we started
+	errCh        chan error            // errCh reports errors encountered while trying to execute task processes
+	completedCh  chan *taskCompletion  // code reports exit codes encountered when task processes exit
+	shouldQuit   chan struct{}         // shouldQuit is closed to indicate that the task should stop its running process, if any
+	done         chan struct{}         // done closes when all processes related to the task have terminated
+	finished     func(ok bool) bool    // callback invoked when a task process has completed; if ok then it will be restarted if it returns true
+	restartDelay time.Duration         // interval between repeated task restarts
+	startProcess taskStateFn           // prepare and start a new live process, defaults to notStartedTask; should be set by run()
+}
+
+// makeTask builds a newly initialized task object but does not start any processes for it
+func makeTask(name string, bin string, args []string, env []string, cl func() io.WriteCloser) *task {
+	return &task{
+		name:         name,
+		bin:          bin,
+		args:         args,
+		env:          env,
+		createLogger: cl,
+		completedCh:  make(chan *taskCompletion),
+		errCh:        make(chan error),
+		shouldQuit:   make(chan struct{}),
+		done:         make(chan struct{}),
+		restartDelay: defaultTaskRestartDelay,
+	}
+}
+
+// spawnTask creates a new task and immediately attempts to spawn a process for the task
+func spawnTask(name string, bin string, args []string, env []string, cl func() io.WriteCloser) *task {
+	t := makeTask(name, bin, args, env, cl)
+	go t.run(notStartedTask)
+	return t
+}
+
+// run executes the state machine responsible for starting, monitoring, and possibly restarting
+// a system process for the task. The initialState func is the entry point of the state machine.
+// Upon returning the done, completedCh, and errCh chans are all closed.
+func (t *task) run(initialState taskStateFn) {
+	t.startProcess = initialState
+
+	defer close(t.done)
+	defer close(t.completedCh)
+	defer close(t.errCh)
+
+	state := initialState
+	for state != nil {
+		next := state(t)
+		state = next
+	}
+}
+
+func (t *task) tryError(err error) {
+	select {
+	case <-t.shouldQuit:
+		// best effort
+		select {
+		case t.errCh <- err:
+		default:
+		}
+	case t.errCh <- err:
+	}
+}
+
+func (t *task) tryComplete(tc *taskCompletion) {
+	select {
+	case <-t.shouldQuit:
+		// best effort
+		select {
+		case t.completedCh <- tc:
+		default:
+		}
+	case t.completedCh <- tc:
+	}
+}
+
+type taskStateFn func(*task) taskStateFn
+
+func taskShouldRestart(t *task) taskStateFn {
+	select {
+	case <-t.shouldQuit:
```

Comment:
:+1: 

initial thinking was that we should make our best effort to terminate here if needed because not doing so could be more expensive later. that said, this could be restructured.